text/0007-gas-price-adjustment.md

```diff
@@ -0,0 +1,135 @@
+- `MIN_FINALIZED_BLOCKS`: The minimum number of blocks that must be finalized in an adjustment period. Otherwise, a fail-safe is triggered. Value: TBD.
+
+### Transaction Fields
+
+- `PRICE_PREMIUM`: The price premium specified for a given transaction.
+- `MAX_GAS_PRICE`: The maximum gas price which the paying signer is willing to accept. The maximum `BASE_PRICE` they are willing to accept is compted as `MAX_GAS_PRICE - PRICE_PREMIUM`.
+
+### Transaction Variables
+
+- `GAS_SPENT`: The total gas spent by a given transaction.
+
+### Adjustment Process
+
+For the `n`-th adjustment period, all finalized blocks with a timestamp in the range `[n*ADJUSTMENT_PERIOD, (n+1)*ADJUSTMENT_PERIOD]` are considered. The total amount of gas used in each one is collected in an array `gas_used_array`. `FULLNESS` for the `n`-th period is computed as:
+
+```python
+FULLNESS = DENOMINATOR * median(gas_used_array) / BLOCK_GAS_LIMIT
+```
+
+Then, a system transaction can be executed in any block proposed after `(n+1)*ADJUSTMENT_PERIOD`, which updates `BASE_PRICE` according to:
+
+```python
+if FULLNESS > TARGET_FULLNESS:
+    new_base_price = BASE_PRICE * (DENOMINATOR + ADJUSTMENT_RATE) / DENOMINATOR
+else:
+    new_base_price = BASE_PRICE * DENOMINATOR / (DENOMINATOR + ADJUSTMENT_RATE)
+
+if new_base_price > INITIAL_BASE_PRICE:
+    BASE_PRICE = new_base_price
+```
+
+This sets the new `BASE_PRICE` for the `n+1`-th adjustment period. The process repeats the same way for every adjustment period.
+
+### Charging for Gas
+
+For each transaction included in a block, the transaction's `MAX_GAS_PRICE - PRICE_PREMIUM` MUST be greater than the block's `BASE_PRICE`. The paying signer will end up paying
+
+```python
+transaction_fee = (BASE_PRICE + PRICE_PREMIUM) * GAS_SPENT
+```
+
+### Fail-safe for Sustained Failure of Finalization
+
+In the event of a sustained finalization failure, where there are too few finalized blocks in an adjustment period, `BASE_PRICE` resets to `INITIAL_BASE_PRICE`:
+
+```python
+if len(gas_used_array) < MIN_FINALIZED_BLOCKS:
+    BASE_PRICE = INITIAL_BASE_PRICE
+```
+
+This is to increase the speed of recovery from a catastrophic failure, by enabling as many users as possible to transact. We choose the value of `INITIAL_BASE_PRICE` low enough, so that resetting practically results in a floating price regime. That is, until the platform recovers from the failure and the algorithm catches up with the market price.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+- The argument for this proposal favors *price stability* over *market efficiency*. In other words, we achieve stability by setting a price floor, and making users occasionally *pay higher* than the actual market price. If this argument is wrong, then the algorithm would be rendered useless.
+- The adjustment rate may fall too short to adapt to short term trends, like sudden increased usage due to a hyped dapp.
```

Comment:
What happens to other applications in this instance?

Reply:
Other dapps? I don't understand the question, but I can try to guess.

If there are 2 dapps, say A and B, where users of A have much higher purchasing power than users of B. (e.g. A = Compound, B = an Ethereum game)

Then if gas price were to increase due to sudden increase in demand by A's users, B's users would be pushed out of the market. From the [same article](https://www.coindesk.com/ethereums-fees-mean-choosing-between-a-world-computer-and-a-financial-network):

> Today, a user sending a multi-million-dollar Tether transaction will most likely outbid someone deploying an Aragon DAO or minting an NFT.

Reply:
Okay got it.  The issue really is around how full blocks are.